### ISSUE TITLE
handling external compilation

### ISSUE THREAD

**Issue by vzaliva** [coq-community/vscoq#768]
Currently, it's challenging to use VsCoq for Coq development involving multiple inter-dependent (.v) modules. Since it doesn't compile them (an issue possibly addressed by https://github.com/coq-community/vscoq/issues/252), one must compile through an external build system like `dune` (as discussed in https://github.com/coq-community/vscoq/issues/134) or `make`.

However, when files are recompiled outside VSCode, it fails to recognize changes in .vo files and continues to display errors based on old definitions. There are a few potential fixes for this:

1. A minimal solution could be to force the reloading of all .vo files when the compilation is done via the Terminal -> Run Build Task menu.

2. A more sophisticated solution would be to monitor .vo files and reload them once they change. This would allow for running `dune build --watch` externally, which would automatically recompile changed files, with VsCoq then picking up these changes automatically as well.

**Comment by vzaliva** [comment]
Sorry, I forgot to mention: I am using VsCoq v2.1.2

**Comment by RalfJung** [comment]
Yeah, I ran into this multiple times within an hour or so of experimenting with vscoq.  I keep forgetting that I not only have to hit `make`, I then also have to hit F1, type "reset", and select that. (I should probably tie that to a shortcut as I will need it a *lot*.)

**Comment by gares** [comment]
We will discuss a roadmap on Monday, so it nice to know this issue concerns many users. (we might make a quick poll also).

The problem is that there are 2 build systems for Coq projects, a dune based one and a make one. The dune based one has a --watch mode where it constantly rebuilds everything and has a RPC to notify when things happen. Maybe we can make it build only the open tabs, but that is not clear yet. Also it is not so clear how maintained the coq dune stanza will be in the future. For the make one there is no builtin facility like that, it seems people home brew it via inotify (on linux) and calls to make. This is ugly.

@RalfJung is your use case a class or some Iris project? I ask because building a student project in --watch mode is probably ok, while building perennial may not really make sense. Also, do you use dune or coq_makefile?

**Comment by RalfJung** [comment]
For now I am only using vscoq in a class -- I'll wait for https://github.com/coq-community/vscoq/issues/252 to happen before using it "in production". So having to type `make` by hand is okay-ish, as the code is small enough that not many things happen cross-file (except for me during preparation of the material, there's plenty of cross-file stuff there, which is why I hit this so often^^).

I'm still using coq_makefile -- sticking to what I know. ;)

If we consider the problem of "how to build dependencies" to be (currently) outside of vscoq -- fixing that is https://github.com/coq-community/vscoq/issues/252 -- the some sort of file watcher does seem like a nice solution. Does vscoq know which .vo files were imported into a given .v file/session? Then it could watch those .vo files and reset the proof state when they change. Not sure how practical that is to do from inside vscode.

**Comment by tchajed** [comment]
I just want to mention that I'm using VS Coq in class _and_ with Perennial. It would still be helpful to have a nice way of building dependencies for the chapters of Software Foundations I assign (everything after Basics.v does depend on compiling) to avoid one set of student confusions.

Building the Perennial dependencies on demand the first time would probably be confusing since it takes a few minutes, and any auto-compilation would need to be aware of this and have a clear indication that compilation is in progress and a way to cancel it.
